modules/internal/ChapelDomain.chpl

```diff
@@ -707,6 +726,8 @@ module ChapelDomain {
 
   // This is the definition of the 'by' operator for domains.
   operator by(a: domain, b) {
+    if ! a.isRectangular() then compilerError(
+      "cannot apply 'by' to associative and sparse domains");
```

Comment:
On the one hand I'd be tempted to make error messages like this say "non rectangular" just on the off chance we end up adding some new domain type later.

On the other hand users may find it clearer if we list it out explicitly and I guess its unlikely to change.

I suppose we could solve this by making the if statement match the error message and adding an assert to catch any unexpected new domain kind:

```
if(a.Sparse() || a.isAssociative) then compilerError(/*what you have now*/)
assert(a.isRectangular());
```

Reply:
Implemented, and the other ones, too.